src/vmm/src/guest_config/x86_64/mod.rs

```diff
@@ -1,17 +1,378 @@
+
+        match template {
+            CpuTemplateType::Custom(template) => {
+                // TODO QUERY ADDITIONAL MSRS
+                let config = Self {
+                    cpuid: supported_cpuid,
+                    msrs: Default::default(),
+                    msr_list,
+                    msr_boot_entries,
+                };
+
+                Self::with_template(config, template)
+            }
+            CpuTemplateType::Static(template) => {
+                // If a template is specified, get the CPUID template, else use `cpuid`.
+                let config_cpuid = match template {
+                    CpuFeaturesTemplate::C3 => cpuid_templates::c3::template(),
+                    CpuFeaturesTemplate::T2 => cpuid_templates::t2::template(),
+                    CpuFeaturesTemplate::T2S => cpuid_templates::t2s::template(),
+                    CpuFeaturesTemplate::T2CL => cpuid_templates::t2cl::template(),
+                    CpuFeaturesTemplate::T2A => cpuid_templates::t2a::template(),
+                    CpuFeaturesTemplate::None | CpuFeaturesTemplate::Custom => {
+                        unreachable!("None and Custom states are invalid")
+                    }
+                };
+
+                // Include leaves from host that are not present in CPUID template.
+                supported_cpuid = config_cpuid
+                    .include_leaves_from(supported_cpuid)
+                    .map_err(Error::CpuidJoin)?;
+
+                // TODO: Some MSRs depend on values of other MSRs. This dependency will need to
+                // be implemented. For now we define known dependencies statically in the CPU
+                // templates.
+
+                // Depending on which CPU template the user selected, we may need to initialize
+                // additional MSRs for boot to correctly enable some CPU features. As stated in
+                // the previous comment, we get from the template a static list of MSRs we need
+                // to save at snapshot as well.
+                // C3, T2 and T2A currently don't have extra MSRs to save/set.
+                match template {
+                    CpuFeaturesTemplate::T2S => {
+                        msr_list.extend(msr_entries_to_save());
+                        cpuid_templates::t2s::update_msr_entries(&mut msr_boot_entries);
+                    }
+                    CpuFeaturesTemplate::T2CL => {
+                        msr_list.extend(msr_entries_to_save());
+                        cpuid_templates::t2cl::update_msr_entries(&mut msr_boot_entries);
+                    }
+                    _ => (),
+                }
+
+                Ok(Self {
+                    cpuid: supported_cpuid,
+                    msrs: Default::default(),
+                    msr_list,
+                    msr_boot_entries,
+                })
+            }
+        }
+    }
+
+    /// Modifies provided config with changes from template
+    fn with_template(
```

Comment:
Just wondering why config is passed in, instead of making this an instance method using &self instead. Do we need an associated function?

Reply:
Fixed